code/game/objects/structures/coathanger.dm

```diff
@@ -1,51 +1,42 @@
 /obj/structure/coatrack
 	name = "coat rack"
-	desc = "Rack that holds coats."
+	desc = "Its a rack that can hold coats, and hats!"
 	icon = 'icons/obj/coatrack.dmi'
-	icon_state = "coatrack0"
+	icon_state = "coatrack"
 	var/obj/item/clothing/suit/coat
-	var/list/allowed = list(/obj/item/clothing/suit/storage/toggle/labcoat, /obj/item/clothing/suit/storage/det_trench)
+	var/obj/item/clothing/head/hat
 
 /obj/structure/coatrack/attack_hand(mob/user as mob)
-	user.visible_message("[user] takes [coat] off \the [src].", "You take [coat] off the \the [src]")
-	if(!user.put_in_active_hand(coat))
-		coat.dropInto(user.loc)
-	coat = null
-	update_icon()
+	if(coat)
+		user.visible_message(SPAN_NOTICE("[user] takes [coat] off \the [src]."), SPAN_NOTICE("You take [coat] off the \the [src]"))
+		if(!user.put_in_active_hand(coat))
+			coat.dropInto(user.loc)
+		coat = null
+		update_icon()
+	else if(hat)
+		user.visible_message(SPAN_NOTICE("[user] takes [hat] off \the [src]."), SPAN_NOTICE("You take [hat] off the \the [src]"))
+		if(!user.put_in_active_hand(hat))
+			hat.dropInto(user.loc)
+		hat = null
+		update_icon()
 
 /obj/structure/coatrack/attackby(obj/item/W as obj, mob/user as mob)
```

Comment:
You never check for if there's already a coat & hat hanging, which would lead to a lot of problems.